src/backend/metal/src/command.rs

```diff
@@ -534,17 +535,101 @@ impl com::RawCommandBuffer<Backend> for CommandBuffer {
 
     fn blit_image<T>(
         &mut self,
-        _src: &native::Image,
+        src: &native::Image,
         _src_layout: ImageLayout,
-        _dst: &native::Image,
+        dst: &native::Image,
         _dst_layout: ImageLayout,
         _filter: com::BlitFilter,
-        _regions: T,
+        regions: T,
     ) where
         T: IntoIterator,
         T::Item: Borrow<com::ImageBlit>
     {
-        unimplemented!()
+        #[inline]
+        fn offset_diff(a: &com::Offset, b: &com::Offset) -> (u64, u64, u64) {
+            let dx = b.x - a.x;
+            let dy = b.y - a.y;
+            let dz = b.z - a.z;
+            debug_assert!(dx >= 0);
+            debug_assert!(dy >= 0);
+            debug_assert!(dz >= 0);
+
+            (
+                dx as u64,
+                dy as u64,
+                dz as u64
+            )
+        }
+
+        #[inline]
+        fn range_size(r: &Range<com::Offset>) -> (u64, u64, u64) {
+            offset_diff(&r.start, &r.end)
+        }
+
+        #[inline]
+        fn is_offset_positive(o: &com::Offset) -> bool {
+            o.x >= 0 && o.y >= 0 && o.z >= 0
+        }
+
+        //TODO we're always switching into blit encoder mode, even when we have no commands to execute on it
+        let encoder = self.encode_blit();
+
+        let blit_cmd = |region: &com::ImageBlit| {
+            let src_start = region.src_bounds.start;
+            let dst_start = region.dst_bounds.start;
+            debug_assert!(is_offset_positive(&src_start));
+            debug_assert!(is_offset_positive(&dst_start));
+
+
+            let src_origin = MTLOrigin {
+                x: src_start.x as u64,
+                y: src_start.y as u64,
+                z: src_start.z as u64
+            };
+            let dst_origin = MTLOrigin {
+                x: dst_start.x as u64,
+                y: dst_start.y as u64,
+                z: dst_start.z as u64
+            };
+
+            let (sx, sy, sz) = range_size(&region.dst_bounds);
+            let src_size = MTLSize {
+                width: sx,
+                height: sy,
+                depth: sz
+            };
+
+
+            debug_assert_eq!(region.src_subresource.layers.len(), region.dst_subresource.layers.len());
+            //TODO aspect flags?
+            for (src_layer, dst_layer) in region.src_subresource.layers.clone().zip(region.dst_subresource.layers.clone()) {
```

Comment:
let's have the `region` argument by move in this closure so that we can avoid `clone()` of the ranges

Reply:
The region argument is borrowed by the whole function, since it takes `IntoIterator<Borrow<com::ImageBlit>>`